### ISSUE TITLE
[Skill] drug-repurposing-screen — general pooled viability screen → repurposing prioritisation

### ISSUE THREAD

**Issue by RezaJF** [ClawBio/ClawBio#278]
## Skill name

`drug-repurposing-screen` (folder: `skills/drug-repurposing-screen/`)

## What it does

Runs a **format-agnostic, objective-driven** end-to-end analysis of **pooled small-molecule viability screens** — from raw plate-level readouts through QC, normalisation, hit calling, optional dose-response fitting, context-selectivity classification, confounder-adjusted biomarker discovery, mechanism checks, and a ranked candidate table.

The skill is **not tied to Corsello et al. 2020 or cancer repurposing**. Users define the repurposing objective in a small YAML config: which sample contexts represent the *target* indication, which represent *off-target* or reference contexts, how compounds are filtered (e.g. approved only, exclude oncology MOA), and which evidence axes matter for prioritisation. The pipeline then asks: *among compounds active in this screen, which show selective activity in the target context with a reproducible molecular signature and translational plausibility for the stated objective?*

**Reference adapter (not the core API):** a bundled `--preset prism-corsello-2020` maps the public [Corsello et al. 2020 PRISM Repurposing bundle](https://doi.org/10.6084/m9.figshare.9393293.v2) onto the generic schema. That preset is the gold-standard integration test and demo-at-scale benchmark; any user-supplied bundle in the same schema works for other objectives (e.g. neurodegeneration models, fibrosis lines, infectious-disease host-cell panels, organoid pools, or in-house repurposing libraries).

The skill fills a gap in ClawBio: existing skills handle **individual pharmacogenomics** (variant → drug response in one genome) but none process **compound × model-panel screening data** (compound → phenotype across many contexts).

## Bioinformatics tools it wraps

- **scipy.optimize.curve_fit** — 4PL Hill dose-response curve fitting (optional stage when multi-dose data exist)
- **NumPy / pandas / scipy.stats** — SSMD, Z-factor, robust median/MAD QC
- **ComBat-equivalent location-scale correction** — batch correction keyed by user-configured batch columns (default: `detection_plate × pool_id`; overridable)
- **sklearn.linear_model.ElasticNetCV + GroupKFold** — predictive model with **user-specified grouping covariate** (default `lineage`; e.g. `disease_subtype`, `batch`, `donor_id` for organoids)
- **statsmodels partial correlation** — confounder-adjusted univariate biomarker sweep
- **SAS bimodality coefficient + Hartigan dip** — context-selectivity classifier
- **Optional dependency phenocopy matrix** — CRISPR/RNAi essentiality or any user-supplied `(sample × gene)` dependency table for mechanism support (DepMap CRISPR is one bundled loader, not a hard requirement)

### Input (any one of)

- **`--bundle <dir> --schema <path>`** — user directory plus a YAML schema that declares:
  - readout file(s): samples × wells matrix (CSV/TSV/parquet)
  - treatment metadata: compound ID, dose, perturbation type (`vehicle_control`, `positive_control`, `trt_cp`, …), plate, well, optional MOA/target/phase columns
  - sample metadata: stable sample ID, optional tissue/lineage/disease labels, pool assignments
  - optional molecular feature matrices: any `(sample × feature)` tables keyed to sample ID
  - optional reference advance list for triage comparison (not required)
  - screen topology: `primary_only`, `secondary_only`, or `two_stage` (primary → dose-response follow-up)
- **`--preset <name>`** — shorthand that loads a known schema + defaults. Shipped presets:
  - `prism-corsello-2020` — Corsello PRISM bundle layout (primary single-dose + secondary 8-point DR, CCLE/DepMap features). *Example repurposing objective out of the box:* non-oncology approved compounds with selective anti-cancer activity.
  - *(future)* community presets for other public screens can be added as adapters without changing the core pipeline.
- **`--objective <path>`** — repurposing objective YAML (required unless preset bundles one). Defines target/off-target sample sets, compound filters, and priority weights.
- **`--demo`** — synthetic two-stage shard (10 samples × 20 compounds, configurable objective: selective activity in `context_A` vs `context_B`). No network required.

### Generic bundle schema (minimum viable)

```
bundle/
  readouts/
    primary.csv          # optional
    secondary.csv        # optional
  metadata/
    treatment_info.csv   # required: maps wells → compound/dose/control/plate
    sample_info.csv      # required: sample IDs + annotation columns
    pooling_info.csv     # optional: pool/barcode assignments
  features/              # optional: any number of sample × feature matrices
    *.csv
  reference/             # optional
    advance_list.csv     # compounds advanced to follow-up (for triage benchmarking)
schema.yaml              # column renames, control-well tags, batch keys, ID canonicalisation rules
objective.yaml           # repurposing goal (see below)
```

### Repurposing objective config (`objective.yaml`)

```yaml
name: "Approved anti-inflammatory → IBD organoid context"   # free-text label for the report
compound_filters:
  clinical_phase: ["Launched", "Phase 2", "Phase 3"]
  exclude_moa: []                    # e.g. ["cytotoxic chemotherapy"] if desired
  require_annotations: ["moa"]       # optional metadata requirements
target_context:
  sample_info_query: "disease == 'IBD' and model_type == 'organoid'"
off_target_context:
  sample_info_query: "disease != 'IBD'"   # used for selectivity framing, not hard exclusion
selectivity:
  prefer: "context_dependent"        # vs "potency_only" or "broad_activity"
priority_weights:
  selectivity: 0.25
  biomarker_strength: 0.25
  clinical_phase: 0.20
  mechanism_novelty: 0.15
  phenocopy_support: 0.15
confounders:
  partial_correlation: ["lineage"]   # columns to adjust in biomarker sweep
  cv_group: "lineage"                # GroupKFold grouping column
```

Users can express **any** repurposing hypothesis by changing `objective.yaml` and sample/compound metadata — not only "non-oncology drugs killing cancer lines."

### Output (`<output_dir>/`)

- `report.md` + `report.html` — narrative report framed around the stated objective (not a fixed cancer-repurposing storyline)
- `result.json` — machine-readable priority table + objective metadata for chaining
- `tables/priority_table.csv` — ranked candidates (`compound_id`, `name`, annotations from user metadata, `priority`, `selectivity_class`, `top_biomarker`, `best_feature_type`, `mechanism_supported`, `objective_name`)
- `tables/selectivity.csv`, `tables/biomarker_univariate_all_matrices.csv`, `tables/dose_response_table_qc.csv` (when DR stage present)
- `figures/` — QC, hit landscape, selectivity, biomarker winners, CV leakage gap, top-N card (labels reflect user objective)
- `cache/` — parquet intermediates; `--resume` for incremental re-runs
- `reproducibility/commands.sh` + `environment.yml` + copy of `schema.yaml` + `objective.yaml`

## Why it matters

- **For screening labs**: collapses ~10 manual stages (QC → prioritisation) into one CLI call for **any** pooled viability repurposing campaign, not only oncology cell-line panels.
- **For translational researchers**: answers *"which compounds in this library merit follow-up for **my** indication?"* with selectivity evidence, confounder-adjusted biomarkers, and configurable clinical/MOA filters — not a single hard-coded cancer narrative.
- **For ClawBio orchestrator**: emits a clean `priority_table.csv` chainable to `target-validation-scorer`, `clinical-trial-finder`, `pubmed-summariser`, and `omics-target-evidence-mapper`.
- **Gap analysis**: no existing ClawBio skill handles compound-by-sample screening data. This generalises the PRISM workflow into a reusable screening skill with public-dataset presets.

## Trigger keywords (proposed)

`drug repurposing`, `repurposing screen`, `viability screen`, `pooled screen`, `compound panel`, `dose-response screen`, `selective activity biomarker`, `repurposing candidate prioritisation`, `PRISM`, `Corsello repurposing hub`, `cell line drug sensitivity`, `organoid drug screen`

**Do NOT fire when:**

- The user asks about *individual* pharmacogenomics (route to `pharmgx-reporter`)
- The user asks about *clinical trial* feasibility for one drug without screening data (route to `clinical-trial-finder`)
- The user asks about *target ID from structure alone* (route to `struct-predictor` or `omics-target-evidence-mapper`)
- The user has only GWAS hits, no compound × sample matrix (route to `gwas-lookup` / `target-validation-scorer`)

## Proposed workflow (generic; preset fills defaults)

1. **Validate** — load `schema.yaml` + `objective.yaml`; confirm readout dimensions, control-well tags, sample ID canonicalisation rules, optional feature matrix joins (≥90% retention floor; halt on ID-system mismatch)
2. **QC** — SSMD (or configured metric) between negative and positive controls per `(sample, batch_unit)`; flag failing pairs and annotation groups
3. **Normalisation + hit calling** — log transform (if configured), fold-change vs per-plate vehicle control, robust z-score, multi-sample hit rule (thresholds CLI-tunable; defaults cite literature, not Corsello-only)
4. **Hit triage** — joint criteria (strength, reproducibility, selectivity-or-potency); **optional** comparison to `reference/advance_list.csv` when provided
5. **Secondary stage** (if present) — inert-barcode/spike-in normalisation when schema declares it; ComBat-style batch correction on configured batch keys
6. **Dose-response fitting** (if multi-dose) — 4PL Hill, IC50 censoring, quality classes (`good` / `low_r2` / `right_censored` / `failed`); gate downstream on `good` curves
7. **Selectivity classifier** — kill rate × median viability × bimodality → `{inactive, broadly_active, context_selective, other}` relative to **target vs off-target contexts from objective.yaml**
8. **Confounder-adjusted biomarker sweep** — all supplied feature matrices; per-compound winner by smallest q-value after adjusting for configured covariates
9. **Mechanism check** — match biomarker to compound annotations (target/MOA/pathway dictionary user-extensible); optional phenocopy via supplied dependency matrix
10. **Composite priority** — weighted evidence axes from `objective.yaml` → ranked top-N **for the stated repurposing goal**

## Chaining partners

- `target-validation-scorer` — `(compound, top_biomarker_feature)` plausibility
- `clinical-trial-finder` — top-N Launched compounds → trial landscape for the **target indication** named in objective
- `pubmed-summariser` — `(compound, biomarker)` literature brief
- `omics-target-evidence-mapper` — multi-omics evidence stack for top-N
- `methylation-clock` / `rnaseq-de` — when biomarker is methylation- or expression-based and user wants follow-up in their own sample

## Demo data plan

Ship under `skills/drug-repurposing-screen/demo/`:

- Generic synthetic bundle matching `schema.yaml` (two contexts: `context_A` selective hits, `context_B` off-target)
- `demo_objective.yaml` — example non-cancer objective (*"Compound X selective in context_A organoids vs fibroblast reference panel"*)
- `demo_features.csv` — 10 samples × 50 features
- Expected: 1–3 context-selective compounds, one biomarker winner, full artefact tree in **<60 s**

**Integration benchmark (not demo):** `--preset prism-corsello-2020 --objective presets/prism_non_oncology_to_cancer.yaml` on the full Figshare bundle reproduces the validated nine-step analysis (reference implementation; ≤15 min laptop with cache).

## Safety notes

- **Local-first** — runs offline once bundle + features are on disk
- **Disclaimer** — ClawBio research-only disclaimer; *candidates are hypotheses for wet-lab follow-up, not clinical recommendations*
- **No hardcoded clinical thresholds** — all cutoffs CLI-tunable; defaults documented with citations
- **Objective-aware reporting** — report language follows `objective.name`, not an assumed oncology repurposing frame
- **Reproducibility bundle** — writes commands, environment, schema, and objective copies

## Gotchas (generic + PRISM preset notes)

1. **Sample ID canonicalisation**: feature matrices and readouts often use different ID systems. Schema must declare a canonical key and mapping rules; silent drops are forbidden (retention reporting + halt below floor).
2. **Batch effects are preset-specific**: e.g. PRISM PMTS003 plates fail SSMD at high rates — handled in `prism-corsello-2020` preset as *drop batch as unit*, not a global hard-code. Other screens may have different batch hazard columns.
3. **Low-R² is usually biology**: most compound × sample pairs are inactive; classify and gate rather than "fixing" the optimiser.
4. **Grouping leakage**: default `GroupKFold` on `objective.confounders.cv_group`; always report random-k-fold gap.
5. **Feature-type winners vary by objective**: methylation may dominate for one repurposing goal, expression for another — report per run, no fixed narrative.
6. **Repurposing objective must be explicit**: without `objective.yaml`, the skill refuses to rank candidates (prevents implicit cancer-only defaults).

## Reference implementation

The validated pipeline behind this proposal (QC → prioritisation, eval harness, method notes) lives in a standalone analysis repo derived from the LatchBio PRISM take-home. The ClawBio skill **ports the logic into the generic schema above**; Corsello/PRISM becomes one preset + integration test, not the product definition.

Key references:

- [Corsello et al. 2020, *Nature Cancer*](https://www.nature.com/articles/s43018-019-0018-6) — reference preset dataset
- [DepMap Repurposing Portal](https://depmap.org/repurposing)
- [Figshare bundle](https://doi.org/10.6084/m9.figshare.9393293.v2)

## Are you willing to implement it?

- [x] Yes, I'd like to build this skill (no committed timeline)

## Acceptance criteria

- [ ] `python skills/drug-repurposing-screen/drug_repurposing_screen.py --demo --output /tmp/drs_demo` completes in <60 s and writes the artefact tree
- [ ] `python skills/drug-repurposing-screen/drug_repurposing_screen.py --preset prism-corsello-2020 --bundle <Corsello dir> --objective presets/prism_non_oncology_to_cancer.yaml --output <dir>` runs end-to-end in ≤15 min on a laptop with cache resume *(reference benchmark, not the only supported input)*
- [ ] A **second objective** on the same Corsello bundle (e.g. `--objective examples/lineage_specific_killing.yaml`) produces a different priority table without code changes — proves generalisation
- [ ] User-supplied generic bundle (documented minimal schema) runs with `--schema` + `--objective` on demo-scale synthetic data
- [ ] SKILL.md passes the 17-point conformance checklist in `CONTRIBUTING.md`
- [ ] `pytest skills/drug-repurposing-screen/tests/` passes (TDD bootstrapped; includes schema validation + objective parsing tests)
- [ ] Catalog regenerated: `python scripts/generate_catalog.py`
- [ ] Routing entry added to `CLAUDE.md` skill-routing table
